src/jvmMain/kotlin/com/charleskorn/kaml/YamlNodeReader.kt

```diff
@@ -194,6 +198,37 @@ internal actual class YamlNodeReader(
         return resolvedNode.withPath(path.withAliasReference(anchor.value, event.location).withAliasDefinition(anchor.value, resolvedNode.location))
     }
 
+    private fun ensureNoDuplicateKeys(node: YamlNode) {
+        accumulateDuplicateKeys(listOf(node))
+            .takeIf { it.isNotEmpty() }
+            ?.mapValues { (_, duplicates) -> duplicates.sortedWith(compareBy({ it.endLocation.line }, { it.endLocation.column }))  }
+            ?.toSortedMap(compareBy { it.endLocation.line })
+            ?.let { throw DuplicateKeysException(it) }
+    }
+
+    private tailrec fun accumulateDuplicateKeys(
```

Comment:
This function and iteration over all nodes could be avoided if something like `YamlNodeFactory` and make it its responsibility to calculate duplicates (and possibly do other stuff for different node types). However, I didn't want to change too much in the code structure.